pkg/frontend/encryptionathost_validation.go

```diff
@@ -0,0 +1,71 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armfeatures"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/features"
+)
+
+type EncryptionAtHostValidator interface {
+	ValidateEncryptionAtHost(ctx context.Context, subscriptionID string, oc *api.OpenShiftCluster) error
+}
+
+type encryptionAtHostValidator struct{}
+
+func (e encryptionAtHostValidator) ValidateEncryptionAtHost(ctx context.Context, subscriptionID string, oc *api.OpenShiftCluster) error {
+	credential, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return err
+	}
+
+	subFeatureRegistrationsClient, err := armfeatures.NewSubscriptionFeatureRegistrationsClient(subscriptionID, credential, nil)
+	if err != nil {
+		return err
+	}
+	return validateEncryptionAtHost(ctx, subFeatureRegistrationsClient, oc)
+}
+
+func validateEncryptionAtHost(ctx context.Context, subFeatureRegistrationsClient features.SubscriptionFeatureRegistrationsClient, oc *api.OpenShiftCluster) error {
+	var hasEncryptionAtHostEnabled bool
+	profilesToCheck := append([]api.WorkerProfile{{EncryptionAtHost: oc.Properties.MasterProfile.EncryptionAtHost}}, oc.Properties.WorkerProfiles...)
+	for _, profile := range profilesToCheck {
+		if profile.EncryptionAtHost == api.EncryptionAtHostEnabled {
+			hasEncryptionAtHostEnabled = true
+			break
+		}
+	}
+	if hasEncryptionAtHostEnabled {
+		err := IsRegisteredForEncryptionAtHostFeature(ctx, subFeatureRegistrationsClient)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func IsRegisteredForEncryptionAtHostFeature(ctx context.Context, subFeatureRegistrationsClient features.SubscriptionFeatureRegistrationsClient) error {
+	response, err := subFeatureRegistrationsClient.Get(ctx, "Microsoft.Compute", "EncryptionAtHost", nil)
+	if err != nil {
+		return err
+	}
+	if *response.Properties.State == armfeatures.SubscriptionFeatureRegistrationStateRegistered {
```

Comment:
Just to make sure we don't get any surprise nil pointer dereferences, let's check for nil pointers as needed before evaluating `*response.Properties.State`.

Reply:
In the new commit, since we are returning nil by default it shouldn't cause any nil pointer dereferences. Let me know your thoughts.

Reply:
I see that you flipped the conditional in response to @cadenmarchese 's comment, but that won't stop us from potentially encountering a nil pointer dereference here. I would still check for nil pointers before trying `if *response.Properties.State ...`.